### PR TITLE
Narrow down non-const enity name expressions within element access expressions

### PR DESCRIPTION
**Pull request by Andarist**
an experiment to fix https://github.com/microsoft/TypeScript/issues/51368 . I wonder what the perf tests show for this (cc @jakebailey ) 😉

### CODE REVIEW

**Comment by fatcerberus** [comment]
To be clear, this *isn't* a general fix for https://github.com/microsoft/TypeScript/issues/56389, correct?

**Comment by Andarist** [comment]
Yeah, it isn't. I specifically only looked into what @DetachHead mentioned [here](https://github.com/microsoft/TypeScript/issues/10530#issuecomment-1809167377)

**Comment by Andarist** [comment]
@jakebailey I fixed the `self-check` here. Would you mind running the perf suite? :)

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2834,8 +2834,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     return findAncestor(errorBindingElement, isBindingElement) !== findAncestor(declaration, isBindingElement) ||
                         declaration.pos < errorBindingElement.pos;
                 }
+                const rootDeclaration = getRootDeclaration(declaration) as Declaration;
+                if (rootDeclaration.kind !== SyntaxKind.VariableDeclaration) {
+                    return true;
+                }
```
Comment (Andarist) [inline src/compiler/checker.ts:2840]: TODO: recheck correctness here

**Comment by jakebailey** [comment]
@typescript-bot test top200
@typescript-bot user test this
@typescript-bot run dt

@typescript-bot perf test this
@typescript-bot pack this

**Comment by Andarist** [comment]
superseded by https://github.com/microsoft/TypeScript/pull/57847

**Pull request closed without merging**